### PR TITLE
Implement Myers algorithm for Levenshtein distance calculation

### LINKED ISSUE #11335: Levenshtein slow and high on CPU usage
## Bug Report

Crystal 1.2.0 (2021-10-13)

LLVM: 11.1.0
Default target: x86_64-apple-macosx

Levenshtein is very slow and uses too much CPU when trying to match a large sample over a large text corpus

Screenshot 2021-10-18 at 16 06 58

https://user-images.githubusercontent.com/10400064/137750280-fc1fd225-0a5c-44b6-b2e1-500a8b93198d.mov

it takes up to 5 minutes to find a match

Case: software license classifier

```crystal
license_file = File.open("src/licenses.csv")
sample = File.read("LICENSE")
license_db = CSV.parse(license_file)
license_db.each do |entry|
  puts Levenshtein.distance(sample, entry[1])
end
```

[licenses.csv](https://github.com/crystal-lang/crystal/files/7365700/licenses.csv)

### PR DESCRIPTION
**Pull request by darkstego**
This implements a faster algorithm for calculating Levenshtein distance. Fixes #11335.

The algorithm is 3x faster than the original one for short strings and 10x for long ones.

### CODE REVIEW

**Comment by vlazar** [comment]
@darkstego I've tried [3rd revision](https://gist.github.com/darkstego/b7f512780454c9088cc8b26a8a6af888) it and results  looks suspicious now 😆 

```
Original: ASCII-Size:3  84.00M ( 11.90ns) (± 5.95%)  0.0B/op        fastest
     New: ASCII-Size:3  83.62M ( 11.96ns) (± 6.11%)  0.0B/op   1.00× slower
Original: ASCII-Size:6  56.30M ( 17.76ns) (± 5.75%)  0.0B/op        fastest
     New: ASCII-Size:6  54.48M ( 18.36ns) (± 6.83%)  0.0B/op   1.03× slower
Original: ASCII-Size:12  81.12M ( 12.33ns) (± 6.54%)  0.0B/op   1.00× slower
     New: ASCII-Size:12  81.46M ( 12.28ns) (± 6.00%)  0.0B/op        fastest
Original: ASCII-Size:24  66.20M ( 15.11ns) (± 5.56%)  0.0B/op        fastest
     New: ASCII-Size:24  65.12M ( 15.36ns) (± 6.52%)  0.0B/op   1.02× slower
Original: ASCII-Size:50  48.46M ( 20.64ns) (± 6.71%)  0.0B/op        fastest
     New: ASCII-Size:50  47.85M ( 20.90ns) (± 5.60%)  0.0B/op   1.01× slower
Original: ASCII-Size:100  34.63M ( 28.88ns) (± 4.48%)  0.0B/op        fastest
     New: ASCII-Size:100  34.11M ( 29.32ns) (± 4.90%)  0.0B/op   1.02× slower
Original: ASCII-Size:500  10.51M ( 95.18ns) (± 3.72%)  0.0B/op        fastest
     New: ASCII-Size:500  10.46M ( 95.60ns) (± 3.80%)  0.0B/op   1.00× slower
Original: ASCII-Size:1000   6.31M (158.47ns) (± 4.93%)  0.0B/op        fastest
     New: ASCII-Size:1000   6.31M (158.59ns) (± 5.67%)  0.0B/op   1.00× slower
Original: Unicode-Size:3  82.80M ( 12.08ns) (± 6.39%)  0.0B/op        fastest
     New: Unicode-Size:3  81.51M ( 12.27ns) (± 6.77%)  0.0B/op   1.02× slower
Original: Unicode-Size:6  55.49M ( 18.02ns) (± 6.11%)  0.0B/op        fastest
     New: Unicode-Size:6  54.78M ( 18.25ns) (± 6.07%)  0.0B/op   1.01× slower
...
```

File: src/levenshtein.cr [hunk 1]
```diff
@@ -10,60 +12,36 @@ module Levenshtein
   # Levenshtein.distance("こんにちは", "こんちは")           # => 1
   # Levenshtein.distance("hey", "hey")              # => 0
   # ```
-  def self.distance(string1 : String, string2 : String) : Int32
+  # If *cutoff* is given then the method is allowed to end once the loweset
+  # possible bound is greater than *cutoff* and return that lower bound.
```
Comment (asterite) [inline src/levenshtein.cr:16]: Could a doc/code example for this be given?

Alternatively, I think it's totally fine to add the `cutoff` logic in a separate PR. Given that it's a new feature, it could sparkle more discussion, eventually leading to this PR being merged less likely or less faster to happen (just a guess! I'm not actually requesting this)

Comment (asterite) [reply src/levenshtein.cr:16]: Good point. I don't think so, but also, the other PR can come/appear after this one is merged.

**Comment by darkstego** [comment]
> @darkstego I've tried [3rd revision](https://gist.github.com/darkstego/b7f512780454c9088cc8b26a8a6af888) it and results looks suspicious now laughing

@vlazar Sorry about that. It was benchmarking two identical strings all the time. Fixed and tested. I also have [a test](https://gist.github.com/darkstego/edc06ddf07e22a983c81c1816c3ab2c7) to compare the distance results between the old and new methods to make sure the data is correct. 

Is there a way to import a module from a file by a different name? There is a lot of copying and pasting that needs to be done to test and benchmark and I would ideally like to test the module that is in my git repo without having to move it around?

File: src/levenshtein.cr (hunk 1 above)
Comment (darkstego) [reply src/levenshtein.cr:16]: I could also make a new pull request and first squash all the features together. Whatever is easier for you guys.

**Comment by vlazar** [comment]
@darkstego Looks great!

New results from me. All cases are now faster, except negligible slowdown on short Unicode strings fro your benchmarks for different string lengths.

1. Comparing to NodeJS on [this benchmark](https://github.com/crystal-lang/crystal/issues/11335#issuecomment-947410609) from original report #11335

Crystal (now 1.6x faster than NodeJS from previously being 8x slower)
```
Time (mean ± σ):      4.404 s ±  0.084 s    [User: 4.364 s, System: 0.019 s]
Range (min … max):    4.227 s …  4.505 s    10 runs
```

NodeJS
```
Time (mean ± σ):      7.269 s ±  0.062 s    [User: 7.133 s, System: 0.183 s]
Range (min … max):    7.163 s …  7.335 s    10 runs
```

2. Benchmark from https://github.com/crystal-lang/crystal/pull/11370#issuecomment-952428372 (short and long strings, small and big distances)

```
old 296.80  (  3.37ms) (± 2.34%)  195kB/op   2.63× slower
new 781.55  (  1.28ms) (± 3.19%)  244kB/op        fastest
```

3. Your benchmark from https://gist.github.com/darkstego/b7f512780454c9088cc8b26a8a6af888

```
Original: ASCII-Size:3  13.98M ( 71.55ns) (± 1.22%)  32.0B/op   1.56× slower
     New: ASCII-Size:3  21.76M ( 45.95ns) (± 4.73%)   0.0B/op        fastest
Original: ASCII-Size:6   5.46M (183.25ns) (± 3.98%)  32.0B/op   2.66× slower
     New: ASCII-Size:6  14.53M ( 68.80ns) (± 4.53%)   0.0B/op        fastest
Original: ASCII-Size:12   1.81M (552.07ns) (± 3.60%)  64.0B/op   4.80× slower
     New: ASCII-Size:12   8.69M (115.12ns) (± 4.06%)   0.0B/op        fastest
Original: ASCII-Size:32 309.26k (  3.23µs) (± 2.91%)   144B/op   7.97× slower
     New: ASCII-Size:32   2.47M (405.65ns) (± 3.26%)  32.0B/op        fastest
Original: ASCII-Size:64  83.17k ( 12.02µs) (± 3.13%)   272B/op  16.74× slower
     New: ASCII-Size:64   1.39M (718.04ns) (± 3.50%)  32.0B/op        fastest
Original: ASCII-Size:100  36.12k ( 27.69µs) (± 3.39%)   448B/op  15.08× slower
     New: ASCII-Size:100 544.75k (  1.84µs) (± 2.76%)  64.0B/op        fastest
Original: ASCII-Size:500   1.60k (623.85µs) (± 5.44%)  2.0kB/op  20.85× slower
     New: ASCII-Size:500  33.43k ( 29.91µs) (± 5.14%)   160B/op        fastest
Original: ASCII-Size:1000 413.47  (  2.42ms) (± 3.95%)  3.93kB/op  21.37× slower
     New: ASCII-Size:1000   8.84k (113.18µs) (± 3.17%)    288B/op        fastest
Original: Unicode-Size:3   6.04M (165.61ns) (± 1.35%)  80.0B/op        fastest
     New: Unicode-Size:3   5.67M (176.31ns) (± 0.95%)  80.0B/op   1.06× slower
Original: Unicode-Size:6   3.39M (294.95ns) (± 1.40%)  96.0B/op        fastest
     New: Unicode-Size:6   3.25M (307.54ns) (± 0.93%)  96.0B/op   1.04× slower
Original: Unicode-Size:12   1.20M (830.25ns) (± 4.47%)  160B/op        fastest
     New: Unicode-Size:12   1.19M (837.91ns) (± 5.32%)  160B/op   1.01× slower
Original: Unicode-Size:32 269.84k (  3.71µs) (± 3.04%)  320B/op        fastest
     New: Unicode-Size:32 269.68k (  3.71µs) (± 3.21%)  320B/op   1.00× slower
Original: Unicode-Size:64  77.03k ( 12.98µs) (± 2.94%)    576B/op   3.31× slower
     New: Unicode-Size:64 254.94k (  3.92µs) (± 0.70%)  1.86kB/op        fastest
Original: Unicode-Size:100  33.37k ( 29.97µs) (± 3.51%)    928B/op   3.96× slower
     New: Unicode-Size:100 132.29k (  7.56µs) (± 0.67%)  2.06kB/op        fastest
Original: Unicode-Size:500   1.52k (655.94µs) (± 3.33%)  4.03kB/op   7.06× slower
     New: Unicode-Size:500  10.76k ( 92.96µs) (± 3.17%)  3.72kB/op        fastest
Original: Unicode-Size:1000 384.41  (  2.60ms) (± 5.13%)  7.87kB/op   7.94× slower
     New: Unicode-Size:1000   3.05k (327.82µs) (± 3.60%)  5.76kB/op        fastest
Total Errors = 0
```

File: src/levenshtein.cr [hunk 2]
```diff
@@ -155,4 +147,194 @@ module Levenshtein
...
+      vp = hnx | ~(d0 | hpx | one)
+      vn = d0 & (hpx | one)
+    end
+    score
+  end
+
+  # Myers Algorithm for ASCII and Unicode
+  #
+  # The algorithm uses uses a dictionary to store string char location as bits
+  # ASCII implementation uses StaticArray while for full Unicode a Hash is used
```
Comment (j8r) [inline src/levenshtein.cr:248]: The explanation is quite good, but why using a Hash for Unicode?

Comment (darkstego) [reply src/levenshtein.cr:248]: Briefly, the dictionary maps char to bit-vector (uint32 or uint64). So a hash is used in the general case. For ASCII chars the codepoint is between 0-128, so for a performance boost we can use a StaticArray of size 128 and use the char codepoint to point to the array address containing the appropriate bit-vector. If we used Arrays and addresses for the entire unicode range it will have to be an array of size 1,114,112.

When I started I tried making a StaticArray of that size, but it was just not working (I guess trying to put something that large on the stack was causing issues). I did try using a regular Array, but from early testing the Hash was not only much smaller memory usage but it had better performance, so I stuck with a Hash.

Comment (j8r) [reply src/levenshtein.cr:248]: Thanks! The issue with StaticArray is related to https://github.com/crystal-lang/crystal/issues/2485. May not make any major difference though. I was mainly wondering about not using Array.

**Comment by paulocoghi** [comment]
Anything holding this PR?

**Comment by beta-ziliani** [comment]
Sorry for the very long delay @darkstego and thanks for the awesome work! It would be great to add a few test cases for longer strings, to make sure we are covering all the different algorithms that are implemented. I can take care of that if you prefer. 🙇

**Comment by darkstego** [comment]
I was actually thinking about moving this into a shard a while back, since I didn't know if it will be mainlined. The algorithm is much more efficient at the cost of more complex code. 

@beta-ziliani I will see if I can track down some of the test code I had. Since the algorithm had a window size (32 or 64 depending on the architecture), it was important to test strings that were multiple window size in length as well as edge cases when the string is exactly equal to the window size (64 in my case). I believe I had a program that generated strings of random lengths and compared the results from my code to the base implementation just to be sure.

**Comment by paulocoghi** [comment]
> I was actually thinking about moving this into a shard a while back, since I didn't know if it will be mainlined. The algorithm is much more efficient at the cost of more complex code.

I vote to mainline the code

**Comment by beta-ziliani** [comment]
In terms of tests, the idea is not to have anything fancy, just a couple of examples with different sizes will do.

**Comment by darkstego** [comment]
Added tests for the Levenshtein implementation. Long unicode and 32,64,70 length ASCII to cover different sizes and all edge cases.

**Comment by paulocoghi** [comment]
Thanks a lot, @darkstego ! I wish your PR gets merged soon by the Crystal team. Excellent contribution!

**Comment by paulocoghi** [comment]
With maximum respect to the Crystal team, it seems that this PR fulfills all the requirements presented with praise.

The PR's author added the last requirements more than one month ago.

File: src/levenshtein.cr [hunk 3]
```diff
@@ -155,4 +147,194 @@ module Levenshtein
...
+
+    if string1.single_byte_optimizable? && string2.single_byte_optimizable?
+      s = string1.to_unsafe
+      l = string2.to_unsafe
+
+      l_size.times do |i|
+        last_cost = i + 1
+
+        s_size.times do |j|
+          sub_cost = l[i] == s[j] ? 0 : 1
```
Comment (HertzDevil) [inline src/levenshtein.cr:167]: `single_byte_optimizable?` means bytes higher than 0x7F should all compare equal, since they behave like the replacement character:

```crystal
Levenshtein.distance("\x81", "\x82").should eq(0)
Levenshtein.distance("\x82", "\x81").should eq(0)
Levenshtein.distance("\x81" * 33, String.new(Bytes.new(33) { |i| 0x80_u8 + i })).should eq(0)

# okay, as these use the `Char::Reader` overload instead
Levenshtein.distance("\x81", "\uFFFD").should eq(0)
Levenshtein.distance("\uFFFD", "\x81").should eq(0)
```

```suggestion
          l_i = {l[i], 0x80_u8}.min
          s_j = {s[j], 0x80_u8}.min
          sub_cost = l_i == s_j ? 0 : 1
```

This also means the `ascii_only?` branches could be relaxed to `single_byte_optimizable?`, operating over an alphabet of 129 "code points" where again bytes larger than 0x7F are mapped to 0x80.

Of course, most strings are already valid UTF-8 so this is a rather low priority issue. Feel free to ignore in this PR.

Comment (darkstego) [reply src/levenshtein.cr:167]: You bring up an interesting point. I am not sure I ever really thought about what the Levenshtein distance of invalid strings would look like. But having said that the I don't even know what the point of this `single_byte_optimizable?` path, because if I am not mistaken no valid string will ever end up there, since if they are ascii it will call a different function and if it isn't then it wouldn't be single_byte_optimizable.

I think past me in an effort to optimize for speed looked into using unsafe pointers in non-ascii strings, to bring a speed boost to that specific scenario. And correct me if I am wrong, isn't this segment of code just providing a faster way to measure the Levenshtein distance of invalid strings? If that is the case then why even have it? 

Maybe getting rid of that 'if block' would be the cleanest solution.

Comment (HertzDevil) [reply src/levenshtein.cr:167]: A string that consists only of ASCII characters plus invalid UTF-8 byte sequences, but not valid ones of 2 bytes or more (code point 0x80 or above), is `single_byte_optimizable?` but not `ascii_only?`.

Comment (darkstego) [reply src/levenshtein.cr:167]: I guess my question is how often these occur in the wild. Is speeding up calculation of Levenshtein distance of strings that are `single_byte_optimizable?` but not `ascii_only?` even of value to anyone? Especially since the `Char::Reader` path is plenty fast to begin with, and handles the invalid bytes properly.

**Comment by ysbaddaden** [comment]
AFAICT there aren't anything else do here. Maybe just add the suggestion by @HertzDevil ?
